portable/GCC/ATmegaxxxx/portmacro.h

```diff
@@ -0,0 +1,149 @@
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * http://www.FreeRTOS.org
+ * http://aws.amazon.com/freertos
+ *
+ * 1 tab == 4 spaces!
+*/
+
+#ifndef PORTMACRO_H
+#define PORTMACRO_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/*-----------------------------------------------------------
+ * Port specific definitions.
+ *
+ * The settings in this file configure FreeRTOS correctly for the
+ * given hardware and compiler.
+ *
+ * These settings should not be altered.
+ *-----------------------------------------------------------
+ */
+
+/* Type definitions. */
+#define portCHAR                    char
+#define portFLOAT                   float
+#define portDOUBLE                  double
+#define portLONG                    long
+#define portSHORT                   int
+
+typedef uint8_t StackType_t;
+typedef int8_t BaseType_t;
+typedef uint8_t UBaseType_t;
+
+#if configUSE_16_BIT_TICKS == 1
+    typedef uint16_t TickType_t;
+    #define portMAX_DELAY           ( TickType_t ) 0xffff
+#else
+    typedef uint32_t TickType_t;
+    #define portMAX_DELAY           ( TickType_t ) 0xffffffffUL
+#endif
+/*-----------------------------------------------------------*/
+
+/* Critical section management. */
+
+#define portENTER_CRITICAL()        __asm__ __volatile__ (                          \
+                                        "in __tmp_reg__, __SREG__"        "\n\t"    \
+                                        "cli"                             "\n\t"    \
+                                        "push __tmp_reg__"                "\n\t"    \
+                                        ::: "memory"                                \
+                                        )
+
+
+#define portEXIT_CRITICAL()         __asm__ __volatile__ (                          \
+                                        "pop __tmp_reg__"                 "\n\t"    \
+                                        "out __SREG__, __tmp_reg__"       "\n\t"    \
+                                        ::: "memory"                                \
+                                        )
+
+
+#define portDISABLE_INTERRUPTS()    __asm__ __volatile__ ( "cli" ::: "memory")
+#define portENABLE_INTERRUPTS()     __asm__ __volatile__ ( "sei" ::: "memory")
+
+/*-----------------------------------------------------------*/
+
+/* Architecture specifics. */
+
+// System Tick  - Scheduler timer
+// Prefer to use the enhanced Watchdog Timer, but also Timer0 is ok.
+
+#if defined(WDIE) && defined(WDIF)              // Enhanced WDT with interrupt capability
```

Comment:
To 1. that is exactly the purpose of this path. The Enhanced Watchdog Timer (WDT) is available on all commonly used ATmega devices. It is configured simply and can be detected by the availability of the Watchdog Interrupt Enable (WDIE), and Watchdog Interrupt Flag (WDIF) provided by avr-libc. These flags are also configured consistently across all MCUs.

To 2. it becomes more challenging, as the numbering for the 16-bit Timers changes from MCU to MCU, as does their configuration and availability. It would be possible to provide a configuration for the ATmega32 (a legacy product), but this wouldn't be applicable to modern ATmega MCU. I've built a matrix of Timers vs MCU for a number of the common MCU types, but it is relatively complex and, as @yuhui-zheng pointed out, reserved bits are masked differently depending on the MCU too.

Happy to add 16-bit Timers as an example, but probably best to be left to the user to do final configuration.

To 3. I'd seek advice on how to best achieve that outcome. Perhaps reviewer edits?